src/core/auth/wallet-authenticator.ts

```diff
@@ -0,0 +1,184 @@
+ * const payloadWithSignature = await sdk.auth.generate({
+ *   application,
+ *   subject: "0x...",
+ * })
+ *
+ * // Then on the client side, we can sign this payload with the wallet requesting to authenticate
+ * const authenticatedPayload = await sdk.auth.sign(payloadWithSignature)
+ *
+ * // Finally, on the server side, we can verify if this payload is valid
+ * const isValid = await sdk.auth.verify(application, authenticatedPayload)
+ * ```
+ * @public
+ */
+export class WalletAuthenticator extends RPCConnectionHandler {
+  constructor(network: NetworkOrSignerOrProvider, options: SDKOptions) {
+    super(network, options);
+  }
+
+  public async generate(
+    payload: AuthenticationPayloadInput,
+  ): Promise<AuthorizedPayload> {
+    const parsedPayload: FilledAuthenticationPayloadInput =
+      AuthenticationPayloadInputSchema.parse(payload);
+
+    const signer = this.requireSigner();
+    const signerAddress = await signer.getAddress();
+
+    const payloadOutput: AuthenticationPayloadOutput =
+      AuthenticationPayloadOutputSchema.parse({
+        // Add the issuer in application:address format to use on the client side
+        iss: `${parsedPayload.application}:${signerAddress}`,
+        sub: parsedPayload.subject,
+        aud: parsedPayload.audience,
+        nbf: parsedPayload.invalidBefore,
+        exp: parsedPayload.expiresAt,
+      });
+
+    const payloadMessage = JSON.stringify(payloadOutput);
+    const signature = await signer.signMessage(payloadMessage);
+
+    const authorizedPayload = AuthorizedPayloadSchema.parse({
+      payload: payloadOutput,
+      authorizedPayload: signature,
+    });
+
+    return authorizedPayload;
+  }
+
+  public async sign(
+    authorizedPayload: AuthorizedPayload,
+  ): Promise<AuthenticatedPayload> {
+    // Ensure that admin address is the same as the issuer
+    const adminAddress = this.recoverAddress(
+      JSON.stringify(authorizedPayload.payload),
+      authorizedPayload.authorizedPayload,
+    );
+    if (
+      authorizedPayload.payload.iss.split(":")[1]?.toLowerCase() !==
+      adminAddress.toLowerCase()
+    ) {
+      throw Error(
+        "Payload issuer is not the same as the address that authorized the payload.",
+      );
+    }
+
+    const signerAddress = await this.requireSigner().getAddress();
+    if (
+      authorizedPayload.payload.sub.toLowerCase() !==
+      signerAddress.toLowerCase()
+    ) {
+      throw Error(
+        "Payload subject is not the same as the wallet connected to the SDK.",
+      );
+    }
+
+    const message = this.generateMessage(authorizedPayload);
+    const signer = this.requireSigner();
+    const authenticatedPayload = await signer.signMessage(message);
+    const payload = AuthenticatedPayloadSchema.parse({
+      ...authorizedPayload,
+      authenticatedPayload,
+    });
+
+    return payload;
+  }
+
+  public async verify(
+    application: string,
+    authenticatedPayload: AuthenticatedPayload,
+  ): Promise<boolean> {
```

Comment:
wonder if this should return 'why' verification failed. maybe with an enum?

Reply:
I don't know if we want verification to return something besides a true/false value here - reason being that the main concern of the endpoint is usually just to check if the user _is_ or _isnt_ authed, and are more concerned with the actual endpoint functionality - the authentication should just be a simple check.

Feels like returning something like `getClaimIneligibilityReasons` with an array of reasons it failed/single reason it failed could be making this common use less intuitive (no longer checking if `authenticated == true`) + with drop it makes more sense b/c there will commonly be times when users don't meet the conditions to claim - whereas in authentication, this verify would really only fail if the user is an attacker, and has tampered with the payload.

If we people specifically need the reasons that a verification failed, maybe it would be good to expose a separate function that explicitly returns reasons for failure which can be used for that case, and maybe internally w/ the verify function

Reply:
Although, as we add in checks for correct application/endpoint, maybe this would be helpful for debugging those 2 cases